hal/src/i2c.rs

```diff
@@ -0,0 +1,412 @@
+//! # Inter-Integrated Circuit (I2C) bus
+
+use core::convert::TryFrom;
+
+use crate::{
+    embedded_hal::blocking::i2c::{Read, Write, WriteRead},
+    pac::{rcc::ccipr::I2C3SEL_A, I2C1, I2C2, I2C3, RCC},
+    rcc::{sysclk_hz, pclk1_hz},
+};
+
+use embedded_time::{fixed_point::FixedPoint, rate::*};
+
+/// I2C error
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum Error {
+    /// Arbitration loss
+    Arbitration,
+    /// Bus error
+    Bus,
+    /// Bus busy
+    Busy,
+    /// Not Acknowledge received
+    Nack,
+    /* Overrun, // slave mode only
+     * Pec, // SMBUS mode only
+     * Timeout, // SMBUS mode only
+     * Alert, // SMBUS mode only */
+}
+
+/// I2C peripheral operating in master mode
+pub struct I2c1<PINS> {
+    i2c: I2C1,
+    pins: PINS,
+}
+
+pub struct I2c2<PINS> {
+    i2c: I2C2,
+    pins: PINS,
+}
+
+pub struct I2c3<PINS> {
+    i2c: I2C3,
+    pins: PINS,
+}
+
+macro_rules! busy_wait {
+    ($i2c:expr, $flag:ident, $variant:ident) => {
+        loop {
+            let isr = $i2c.isr.read();
+            let icr = &$i2c.icr;
+
+            if isr.arlo().is_lost() {
+                icr.write(|w| w.arlocf().clear());
+                return Err(Error::Arbitration);
+            } else if isr.berr().is_error() {
+                icr.write(|w| w.berrcf().clear());
+                return Err(Error::Bus);
+            } else if isr.nackf().is_nack() {
+                while $i2c.isr.read().stopf().is_no_stop() {}
+                icr.write(|w| w.nackcf().clear());
+                icr.write(|w| w.stopcf().clear());
+                return Err(Error::Nack);
+            } else if isr.$flag().$variant() {
+                break;
+            }
+        }
+    };
+}
+
+macro_rules! i2c {
+    ($($I2cX:ident: ($I2CX:ident, $i2cXen:ident, $i2cXrst:ident, $i2cXsel:ident,
+                     $I2cXSda:ident, $I2cXScl:ident, $i2cXsclAf:ident, $i2cXsdaAf:ident),)+) => {
+        $(
+            impl<SCL, SDA> $I2cX<(SCL, SDA)> {
+                /// Enables clock and resets the peripheral
+                fn enable_clock(rcc: &mut RCC) {
+                    rcc.apb1enr1.modify(|_, w| w.$i2cXen().enabled());
+                    rcc.apb1rstr1.modify(|_, w| w.$i2cXrst().reset());
+                    rcc.apb1rstr1.modify(|_, w| w.$i2cXrst().clear_bit());
+                }
+
+                fn clock(rcc: &RCC) -> Hertz {
+                    // NOTE(unsafe) atomic read with no side effects
+                    match rcc.ccipr.read().$i2cXsel().variant().unwrap() {
+                        I2C3SEL_A::HSI16 => Hertz(16_000_000),
+                        I2C3SEL_A::SYSCLK => Hertz(sysclk_hz(rcc)), // TODO move the HAL to embedded-time?
+                        I2C3SEL_A::PCLK => Hertz(pclk1_hz(rcc)), // TODO Correct this (HCLK1 / DIV APB1)
```

Comment:
Huh, I didn't know this crate existed before now!

I do like the performance benefits it offers, though it will increase binary size when using `defmt` since there are some explicit panics and unwraps (in dependencies).

Does it offer any benefits for this specific use-case?  For the `Hertz` type It looks like it is just a wrapper around a `u32`.

Reply:
I haven't played enough with it, but to me, the benefit is a clear way of thinking about frequencies/timing throughout the whole HAL and not having to deal with Hertz/kHz/MHz/ms/us/etc.

Using the HAL already adds a lot of binary overhead, and as far as I've seen in a few isolated tests, the additional panics don't account for much since defmt just creates references to the messages, AFAIK.

I'm happy to try porting the rest to use this library and comparing binary size & readability in another PR.

Reply:
> the benefit is a clear way of thinking about frequencies/timing throughout the whole HAL and not having to deal with Hertz/kHz/MHz/ms/us/etc in raw form (i.e. as u32/u64).

That is good enough for me, I am happy to keep using it.